File: day24.nim
import strutils, sequtils, math, algorithm

const
  instructions = readFile("./inputs/24.txt").splitLines().map(parseInt)
  firstGoal = instructions.sum() div 3
  secondGoal = instructions.sum() div 4

type Package = tuple
  amount: int
  quantum: int

var presents = sorted(instructions, cmp, SortOrder.Descending)


proc calc(remaining: int, presents: seq[int], combinations: var seq[Package],
          used=0, qe=1) =
  if remaining == 0:
    combinations.add((used, qe))
  elif remaining > 0 and used < 6 and len(presents) > 0:
    let
      first = presents[0]
      rest = presents[1 .. presents.high]
    calc(remaining-first, rest, combinations, used+1, qe*first)
    calc(remaining, rest, combinations, used, qe)

proc findSolution(goal: int): int =
  var combinations = newSeq[Package]()
  calc(goal, presents, combinations)
  combinations.sort(cmp)
  return combinations[0][1]


echo findSolution(firstGoal)
echo findSolution(secondGoal)

File: python/day24.py
def solve(data, groups):
    def aux(presents, score, used, qe):
        if score == goal:
            possible_solutions.add((used, qe))
        elif score < goal and presents and used < 6:
            aux(presents[1:], score, used, qe)
            aux(presents[1:], score+presents[0], used+1, qe*presents[0])

    possible_solutions = set()
    goal = sum(data) // groups
    aux(data, 0, 0, 1)
    return min(possible_solutions)[1]


data = sorted(map(int, open("inputs/24.txt").read().splitlines()), reverse=True)

print(solve(data, 3))
print(solve(data, 4))
